services/github/github-auth-service.js

```diff
@@ -39,7 +41,34 @@ class ConditionalGithubAuthService extends BaseJsonService {
   }
 }
 
+class GithubAuthV4Service extends BaseGraphqlService {
+  constructor(context, config) {
+    super(context, config)
+    this._requestFetcher = createRequestFetcher(context, config)
+    this.staticAuthConfigured = true
+  }
+
+  async _requestGraphql(attrs) {
+    attrs.url = `/graphql`
+
+    /*
+    The Github v4 API requires us to query the rateLimit object to return
+    rate limit info in the query body instead of the headers:
+    https://developer.github.com/v4/guides/resource-limitations/#returning-a-calls-rate-limit-status
+    This appends the relevant rateLimit query clause to each
+    call to the GH v4 API so we can keep track of token usage.
+    */
+    attrs.query = mergeQueries(
+      attrs.query,
+      'query { rateLimit { limit cost remaining resetAt } }'
```

Comment:
Explicitly asking the GH API to just tell us what our remaining usage is in the body seems like an easier approach than attempting to parse the query and calculate the projected cost of each call and less likely to result in a discrepency: https://developer.github.com/v4/guides/resource-limitations/#calculating-nodes-in-a-call

That said, we probably do need to watch out for query count on contributions. With the V3 API, its pretty hard for someone to contribute a badge that smashes our usage limit because 1 http request = 1 point, but with a graphql a single http request can cost you a whole truckload of points. Not sure what the solution is for that other than to watch out for it in review.

Reply:
How about establishing a max number of queries, which could be overridden per service? If the response exceeds the configured number it could error out rather than rendering. That way we'll see it right in the code, and can even search for it.

On the other hand, we have a rather huge amount of GitHub quota – I want to say something like 2.2k tokens – so maybe we don't need to worry about this.

Reply:
Are you thinking of that as a runtime check or a static check? Either way, I'm not against adding it, but I think it can be raised as an issue to address in a later PR.

Reply:
Since we don't know the answer until the response comes back, it'd have to be a runtime check. Agreed, it can be addressed later.

Reply:
Right, ok. I misunderstood and thought you were still talking about pre-parsing the query before we send it.

Throwing the exception after we've already made the query is obviously simpler, but also of more limited value (esp if it makes it through to production) in that we've already made the request we don't want to make before deciding we didn't want to make it, if you see what I mean :)

Maybe its a useful MVP to start with, in that if we start getting exceptions in sentry saying "lol - just rinsed our token and now we're not even going to render a badge with all the expensive data we just requested" we can at least be reactive. Bear in mind it might lull us into a false sense of security though as we'd be unlikely to throw it in dev because we treat global tokens differently from pooled ones. My instinct is most contributors probably use a global token as opposed to a pool with one token in it (I know that's what I usually do - I think working on this PR is the first time I've set up a pool locally).

Unfortunately I think we do still actually need to work out the query cost up-front to implement it in a useful way.